### PR TITLE
Firefly-646: use fits data on clent

### PR DESCRIPTION
**Pull request by robyww**
Firefly-646: use FITS data on client

### After extensive evaluation I did the following.

- load byte (stretch data) is much more feasible, since it is 1/4 the size and then can be compress. Also, it does not overwhelm the client.

- Added web worker and GPU processing
- Added some UI updates:
  - scroll wheel zoom
  - Improved color bar dropdown
  - scale bar color change
  - scale bar compress and bias (like DS9)

#### Overview

- Retries FITS data to client and then all operation happen locally
- Uses WebWorkers to off load some of the processing
- Uses GPU.js to to access the GPU for faster processing.

#### Testing
https://fireflydev.ipac.caltech.edu/firefly-646-fits-workers-gpu/firefly/
- The FITS data is not loaded until you move your mouse over the image (_This is a placeholder for when to load the data_)
- The title label changes color (more white) slightly FITS data is loaded (_this is here to help with testing_)
- All features should be the same after the data is loaded.

#### To Notice:

- On the slow IPAC VPN FITS data loads _much_ slower than expected.
  - if you are testing over the VPN load smaller images or a one bigger image at a time.
- All features are either faster or more consistent.
- Flux readout is immediate.
- Zoom is much, much smother.
- Color change and stretch are about the same when an image is zoomed to fit the screen, but always more consistent.
  - These features can be much faster depending on network and server load
- Images sizes and performance:
   - Chrome performs really well with almost any image size: I have tested up to 750 MB
   - Firefox can handle big images but performance does not match chrome, you can see a slight lag
   - Safari can really handle an image much over 10 or 15 MB.

#### For developer's: 2 new technology additions.

- Web Workers
- GPU.js

### CODE REVIEW

File: src/firefly/js/util/fetch.js [hunk 1]
```diff
@@ -6,13 +6,10 @@ import {getOrCreateWsConn} from '../core/messaging/WebSocketClient';
 import {ServerParams} from '../data/ServerParams';
 import {showInfoPopup} from '../ui/PopupUtil';
 import {DownloadProgress, getDownloadProgress} from '../rpc/SearchServicesJson';
+import {AJAX_REQUEST, doFetchUrl, REQUEST_WITH, WS_CHANNEL_HD, WS_CONNID_HD} from './lowerLevelFetch.js';
```
Comment (loitly) [inline src/firefly/js/util/fetch.js:9]: It has to be in a new `lowerLevelFetch.js` file? Cannot just export `doFetchUrl` from within the same file?

**Review by loitly — APPROVED** [review]

**Comment by loitly** [comment]
This is what I am seeing:
In Chrome, the mouse-over performance gain is most noticeable after `dev-tools` is opened.
In Safari and Firefox, it still pauses while the mouse is moving regardless if `dev-tools` is opened or not.  I do see output from console that `workers` are used.  And, while the mouse is moving, it does flashes of update which is not the case before.  I don't know why having Chrome dev-tools up would make a difference.

File: config/web.xml [hunk 2]
```diff
@@ -78,6 +78,10 @@
         <servlet-name>FireFly CommandService</servlet-name>
         <url-pattern>/sticky/CmdSrv</url-pattern>
     </servlet-mapping>
+    <servlet-mapping>
+        <servlet-name>FireFly CommandService</servlet-name>
+        <url-pattern>/sticky/CmdSrvNoZip</url-pattern>
+    </servlet-mapping>
```
[the remark this answers is not in the document]
Comment (robyww) [reply config/web.xml:84]: Thanks- I will talk to you more about it to understand

File: src/firefly/js/util/fetch.js (hunk 1 above)
Comment (robyww) [reply src/firefly/js/util/fetch.js:9]: @loitly already talked about this, but for the record... `fetch.js` bring too much (react,logger,web socket stuff) to push into a worker. The worker grow to large. The lowLevelFetch.js only brings in fetch.  @loitly has suggested we change this file into some so of lowlevel util file. The might be worth doing, right now `WebUtil.js` mostly serves that purpose. Maybe it should be moved to `WebUtil.js`

File: config/web.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (robyww) [reply config/web.xml:84]: Zipping the float array was a significant performance impact and provided very little benefit.

[the remark this answers is not in the document]
Comment (robyww) [reply config/web.xml:84]: @loi and I have talked about this but for the record:
`Accept-Encoding` is a forbidden header?
https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name

**Pull request merged**